### DISCUSSION TITLE
Avalon Hope

### DISCUSSION THREAD

**Discussion by avalonhope** [evennia/evennia#2766] [Greetings & Introductions]
Hello,

I am developing a sci-fi game based on the Evennia framework and am also interested in contributing to the core Evennia code, probably by helping to write unit tests and fix bugs as needed. I work as a Python developer but do not claim to be an expert yet.  Contributing to Evennia will help me to improve my Python skills.

-Avalon

**Comment by Griatch** [comment]
Hi @avalonhope! Welcome to Evennia! We are always happy to see people start on new projects in Evennia; it's actually not too unusual to see people making a game in Evennia specifically as a way to hone their development skills. 
And of course. both bug fixes and more unit tests are always welcome :grin: 

Tell us more about your sci-fi game, what kind of sci-fi are we talking about? Space opera? Dystopian survival? Cyberpunk? Something else?

**Reply by avalonhope** [reply]
Very much space opera, but my second-favorite genre is medieval fantasy, so I plan to combine the two by having some low-tech worlds with only medieval technology in common use but where advanced technology seems like magic. Also trying to work in the concept of plural systems (multiple personalities) and inner (fantasy) worlds.  I want space travel to be slow but not too slow, so that it takes several weeks to span the whole galaxy and a few days between star system.  Still working out many of the details; for example a starship should contain rooms, yet also be mobile; maybe this can be handled by location tags?

**Reply by Griatch** [reply]
Interesting! What do you mean by multiple personalities in this context? And how would an 'inner (fantasy) world' work game-wise? 

A starship can well be a group of rooms that are represented in the actual game world as a spaceship object moving around the grid - there is no need to have the rooms _actually_ be inside said ship (but they could easily be, technically. 

Exactly how to handle space travel varies; it can be done with coordinates that you then check if they overlap or come closer to another ship (to have an encounter). More abstract systems (like one using jump gates or similar) could even let each star system be a 'room' and it just takes time traversing the exits between them.

**Reply by InspectorCaracal** [reply]
Hi! I'm also working on a "space fantasy" kind of game (though it sounds completely different from what you're working on so don't worry about that xD).

I'm intrigued about you wanting to work in plurality. Are you planning on it being a theme for the worldbuilding, or an actual mechanic? I feel like trying to make plurality into a game mechanic would be really tricky - not so much with implementations as with there being so many different ways it works.

**Reply by avalonhope** [reply]
So far, the game mechanic for plurality works as follows.  A character can use the meditate command to enter a special room that represents their inner world, and the imagine command to create characters located in that room.  Inner worlds are private to that character but there are also shared inner worlds, accessible to characters of the same faction.  In both cases the awaken command returns the character to their previous location in the outer world.  For worldbuilding, there might also be some NPCs with complex inner worlds that players can enter through some kind of telepathy or mindmeld.

I have yet to implement safeguards to prevent characters from dropping physical objects into their inner worlds or from taking objects back out.  I might have to tag items in the inventory.  Also I have yet to implement switching; the character in the outer world is currently the same character in the inner world and the inner world companions/tulpas do not do anything.  Ideally I would like for players to be able to build out their inner worlds and possibility modify the parameters for game mechanics without their inner worlds.

Yes, I was thinking about having star systems and planets as rooms, but also having regions of empty deep space where ships could be parked or stations built.  So, it sounds like rooms can be used on a number of different levels.  A character could be in a room associated with a starship and the starship object could be in a room associated with a star system.  Yes, I think that will work fine.

As well as being able to modify their inner world, I would like to have ways for characters to build out the main universe, by building ships, constructing stations and discovering star systems.  I would like to do a biospheres system a bit like TerraGenesis in which players can create lifeforms such as microbes, plants and animals in order to terraform a planet.

For advanced worlds, I am trying to work out an alternative to money, some kind of in-game AI algorithm that determines if a character is entitled to  use a resource, based on their needs and privileges.  So advanced societies act as a kind of collective or corporation, without individual wealth, but individuals can accumulate reputation or status based on their productivity in creating goods and services as well as their frugality in consumption of resources.  Instead of one person one vote, higher status individuals would have a weighted vote with more influence.  High tech weapons would have AI controls preventing misuse.  Crime would be difficult due to high tech surveillance.

On the other hand, the medieval worlds would be feudal, using barter, coins and letters of credit, yet to be implemented.

I made a little bit of progress, and then I started writing tests, which slowed me down a bit, but will hopefully avoid the need for extensive manually testing in future.

**Comment by Griatch** [comment]
It sounds like it would be easiest to implement your 'inner world' by simply having two different character objects - depending on if you are awake or mediating you just puppet the other one. Your real world character object can just remain in the world while you puppet the 'inner world' character. 

Then you don't have to worry about moving physical objects into your inner world - the two would be completely separate (your puppet-swap would technically have you puppet another character in a different location). 

(With Evennia multisession modes you could even theoretically control both your real and 'inner' character at the same time from different clients. Pretty trippy and probably very confusing, but doable. :grin: )

**Reply by avalonhope** [reply]
Thanks - that sounds like a neater way to implement inner/outer worlds, especially the fact that the inner person may be different to the outer person.
